### PR TITLE
Broaden no_expect lint for path-loaded test modules (incl. nonstandard names)

### LINKED ISSUE #132: `no_expect_outside_tests` does not recognise `#[tokio::test]`
## Summary
`no_expect_outside_tests` flags `.expect(..)` inside `#[tokio::test]` functions
as non‑test code. This happens even when `tokio::test` is listed in
`dylint.toml` under `additional_test_attributes`. The same code passes once the
module is renamed to `tests` (or `mod tests { ... }` is introduced), suggesting
the lint is not consistently detecting async test attributes in non‑`tests`
modules or in modules loaded via `#[path = "..."]`.

## Environment
- Whitaker suite via `whitaker-installer --experimental`
- Toolchain: `nightly-2025-09-18`
- Cargo Dylint via `whitaker --all`
- OS: Linux (x86_64)

### 1. Create a crate

```
cargo new tokio-whitaker-repro --lib
cd tokio-whitaker-repro
```

### 2. Add tokio dev‑dependency

```toml
# Cargo.toml
[dev-dependencies]
tokio = { version = "1", features = ["macros", "rt"] }
```

### 3. Add `dylint.toml`

```toml
# dylint.toml
[no_expect_outside_tests]
additional_test_attributes = ["tokio::test"]
```

### 4. Add a test module via `#[path]` with `#[tokio::test]`

```rust
// src/lib.rs
#[cfg(test)]
#[path = "service_tests.rs"]
mod service_tests;
```

```rust
// src/service_tests.rs
#[tokio::test]
async fn tokio_expect_is_allowed() {
    let value = Ok::<_, ()>("ok");
    value.expect("ok");
}
```

### 5. Run Whitaker

```
whitaker --all
```

## Actual Result

Whitaker reports `no_expect_outside_tests` errors:

```
error: Avoid calling expect on `std::result::Result<..., ...>` outside test-only code.
  --> src/service_tests.rs:3:5
   |
3  |     value.expect("ok");
   |     ^^^^^^^^^^^^^^^^^^^
   |
   = note: The call originates within function `tokio_expect_is_allowed` which is not recognised as a test.
```

## Expected Result

`#[tokio::test]` functions (and other async test frameworks listed in
`additional_test_attributes`) should be recognised as test contexts. The lint
should not emit a diagnostic for `.expect(..)` inside those tests.

## Additional Observations

- Renaming the module to `tests` and loading it as `#[cfg(test)] mod tests;`
  makes the lint accept the same `#[tokio::test]` function.
- The lint source already includes `tokio::test` in its internal test attribute
  list, so the behaviour appears inconsistent with the implementation.

## Suggested Investigation Areas

- HIR attribute handling for `#[tokio::test]` in modules loaded via `#[path]`.
- Whether `ContextEntry` attribution is missing attributes for the function
  items in this path‑loaded module.
- Whether `additional_test_attributes` is parsed but not applied in this case.

### LINKED ISSUE #189: Recognize `rstest` and `rstest_parametrize` as test-only code in Whitaker
## Problem statement

Whitaker currently fails to recognize test functions declared with `#[rstest]` or `#[rstest_parametrize]` as test-only code. That causes false positives in lints that intentionally forbid patterns such as `.expect()` or `.unwrap()` outside tests.

This is a Whitaker bug or gap in test-context detection, not an rstest bug.

## Current behavior

Given code like this:

```Rust
#[rstest]  
#[case::syntactic(LockFailureKind::Syntactic, "syntax error")]  
#[case::semantic(LockFailureKind::Semantic, "type error")]  
fn content_transaction_rejects_lock_failure(  
    #[case] kind: LockFailureKind,  
    #[case] message: &str,  
) {  
    let dir = TempDir::new().expect("temp dir");  
    // ...  
}
```

Whitaker reports `.expect()` as if it were being called outside test-only code.

Example diagnostic:

```
error: Avoid calling expect on `std::result::Result<tempfile::TempDir, std::io::Error>` outside test-only code.  
  --> crates/weaverd/src/safety_harness/transaction/content_transaction_tests.rs:62:15  
   |  
62 |     let dir = TempDir::new().expect("temp dir");  
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^  
   |  
   = note: The call originates within function `content_transaction_rejects_lock_failure` which is not recognised as a test.
```

That diagnosis is wrong. The function is test code.

## Root cause

Whitaker’s current test-detection logic appears to rely on patterns such as:

- `#[test]`
- well-known async test attributes such as `#[tokio::test]`
- possibly naming heuristics such as `test_*`

That misses rstest, because `#[rstest]` and related forms are procedural macro attributes that imply test semantics without necessarily exposing a plain `#[test]` attribute at the AST stage Whitaker inspects.

In other words, Whitaker currently treats “is this syntactically a standard Rust test?” as too close to “is this test-only code?”. Those are not equivalent.

## Expected behavior

Whitaker should classify functions as test-only code when they are annotated with rstest test macros, including at least:

- `#[rstest]`
    
- `#[rstest_parametrize]`
    

It may also be worth considering whether supporting associated rstest patterns such as `#[case::...]` improves robustness, but the primary detection should sit on the function-level attributes.

For the example above, Whitaker should not emit `no_expect_outside_tests` or similar diagnostics.

## Scope

This affects any Whitaker lint that needs to distinguish production code from test-only code, including but not limited to:

- `no_expect_outside_tests`
- `no_unwrap_outside_tests`
- any future lint that grants more latitude inside tests
    

It also matters for projects using:

- `rstest`
- `rstest-bdd`
- other proc-macro-based test frameworks with nonstandard test markers
    

## Why Whitaker should handle this

It is reasonable for lint tools to recognize widely used test frameworks directly. Requiring every upstream test framework to emit custom tooling hints, or requiring every downstream project to configure special cases, creates unnecessary friction.

`rstest` is established enough that first-class support in Whitaker makes sense.

More importantly, the semantic question Whitaker is trying to answer is not “does this function have a literal `#[test]` attribute?” but “is this function test-only code?”. `#[rstest]` clearly satisfies that criterion.

## Workarounds in the meantime

Projects can work around the issue, but all current workarounds are inferior to fixing detection in Whitaker.

### Return `Result` from tests

```Rust
#[rstest]  
#[case::syntactic(LockFailureKind::Syntactic, "syntax error")]  
fn content_transaction_rejects_lock_failure(  
    #[case] kind: LockFailureKind,  
    #[case] message: &str,  
) -> Result<(), String> {  
    let dir = TempDir::new().map_err(|e| format!("temp dir: {e}"))?;  
    // ...  
    Ok(())  
}
```

This avoids `.expect()`, but it changes test style and pushes projects toward ceremony they may not want.

### Add redundant attributes

```Rust
#[test]  
#[rstest]  
#[case::foo(1)]  
fn my_test(#[case] x: i32) { ... }
```

This is clumsy, may be fragile, and should not be necessary merely to satisfy a lint tool.

### Add framework-specific configuration

Whitaker could support configurable test attributes, for example:

```TOML
[no_expect_outside_tests]  
test_attributes = ["test", "rstest", "rstest_parametrize"]
```

That is better than nothing, but rstest is common enough that built-in support is preferable.

## Suggested fix

Extend Whitaker’s existing “is this test-only code?” detection to recognize rstest attributes directly.

At minimum:

- treat a function annotated with `#[rstest]` as test-only
- treat a function annotated with `#[rstest_parametrize]` as test-only
    

If Whitaker already has shared test-detection logic used by multiple lints, this should go there rather than being implemented ad hoc in one lint.

It would also be sensible to structure the detection as a broader framework mechanism, so additional test frameworks can be supported consistently.

## Acceptance criteria

- A function annotated with `#[rstest]` is recognized as test-only code.
- A function annotated with `#[rstest_parametrize]` is recognized as test-only code.
- `no_expect_outside_tests` does not fire for `.expect()` calls inside such functions.
- Existing support for `#[test]` and async test frameworks continues to work.
- The test-detection logic remains reusable across Whitaker lints.

## Example regression test

```Rust
#[rstest]  
#[case(1)]  
fn rstest_allows_expect_in_test_context(#[case] value: i32) {  
    let parsed = Some(value).expect("value should be present");  
    assert_eq!(parsed, 1);  
}
```

Whitaker should not report `no_expect_outside_tests` here.

## Environment

- Whitaker issue observed against rstest 0.26.1
- Rust 1.75+  
- Related Whitaker issue: #188

## Investigation in Whitaker

I reviewed the current test-detection path before assuming the recent Tokio work had dropped `rstest` support.

This does **not** look like a simple matcher regression. The shared test-like attribute registry in `common/src/attributes/mod.rs` still includes:

- `rstest`
- `rstest::rstest`
- `case`
- `rstest::case`

The HIR adapter in `src/hir.rs` still delegates to that shared matcher, and `no_expect_outside_tests` still uses that path in `crates/no_expect_outside_tests/src/driver/mod.rs`.

That suggests the recent Tokio work did not simply remove `rstest` from the recognised attribute list.

### Likely culprit: hollow rstest regression coverage

The more likely problem is that Whitaker’s existing rstest coverage is too weak to catch a real regression.

The positive rstest UI fixture is:

- `crates/no_expect_outside_tests/ui/pass_expect_in_rstest.rs`

That fixture depends on the auxiliary proc-macro:

- `crates/no_expect_outside_tests/ui/auxiliary/rstest.rs`

The auxiliary macro simply prepends a literal `#[test]` to the item. That means the current UI test proves only that “a proc macro which emits `#[test]` is accepted”. It does **not** prove that real `rstest` is recognised correctly by Whitaker.

In other words, if native rstest detection were broken, this test could still pass.

### Contrast with the recent Tokio work

The recent Tokio fix explicitly hardened coverage with a real example-based regression:

- `crates/no_expect_outside_tests/src/lib_ui_tests.rs`
- `crates/no_expect_outside_tests/examples/pass_expect_in_tokio_test_harness.rs`
    

That regression compiles a real `#[tokio::test]` example under `--test`, precisely because the earlier proc-macro fixture only covered token shape rather than the real lowering and harness path.

Whitaker currently lacks an equivalent end-to-end regression for real rstest usage.

## Implication

The safest reading is:

- the recent Tokio work may have exposed the gap, but
- the underlying problem is more likely that Whitaker has a hollow rstest regression test which masks a real detection failure.

So the issue is probably **not** “Tokio support removed rstest from the matcher”. It is more likely “Whitaker does not currently prove that real rstest tests are classified as test-only code”.

## Proposed follow-up

1. Add a real rstest example-based regression using the actual `rstest` crate rather than the current auxiliary proc-macro that injects `#[test]`.
    
2. If that regression fails, inspect the lowered or HIR shape to determine whether rstest needs:
    
    - only direct source-level attribute recognition, or
    - an additional fallback path analogous to the Tokio harness-descriptor recovery.
        
3. Treat the current `pass_expect_in_rstest` fixture as insufficient on its own for future regressions.

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Broaden detection to recognise test contexts inside `#[path]`-loaded modules with non-standard names (e.g., `service_tests`) under the `--test` harness, and to recognise test-like contexts introduced by `rstest`/`rstest_parametrize` attributes. This ensures `Option::expect` lint rules apply in nested path-modules containing `#[tokio::test]`, `#[case]`-generated tests, etc.

## Changes
- Core logic
  - Add `has_test_module_name(name: &str) -> bool` in `crates/no_expect_outside_tests/src/driver/test_context.rs` to recognise test module naming conventions:
    - Exact matches: `test`, `tests`
    - Prefixes: `test_`, `tests_`
    - Suffixes: `_test`, `_tests`
    - Also covers non-standard path-loaded module names such as `service_tests`.
  - Update `is_test_named_module` to delegate to `has_test_module_name` instead of hard-coded checks.
  - Expanded documentation with examples and rationale for non-standard, path-loaded module names, including rstest_parametrize contexts.

- Tests and fixtures
  - Unit tests: Extend `crates/no_expect_outside_tests/src/driver/tests.rs` with `has_test_module_name_matches_test_conventions` using `rstest` to cover a broad set of naming patterns, including `service_tests` and other non-standard forms, and `rstest_parametrize` variants.
  - UI/UI-regression tests: Introduced fixtures to exercise path-loaded modules with non-standard names under `--test`:
    - `crates/no_expect_outside_tests/examples/pass_expect_in_path_module_harness.rs`
    - support module `path_module_harness_support/service_tests.rs` to simulate a path-loaded test module.
    - UI tests for path-loaded tokio tests:
      - `crates/no_expect_outside_tests/ui/pass_expect_in_path_module_tokio_test.rs`
      - accompanying `service_tests.module` to exercise `#[tokio::test]` inside a path-loaded module.
      - `crates/no_expect_outside_tests/ui/pass_expect_in_cfg_test_named_module.rs` and related stderr fixture to verify `#[cfg(test)]`-guarded non-standard module names.
  - Regression UI test: Added `path_module_tokio_test_compiles_under_test_harness` to verify that a path-loaded tokio test compiles under the test harness (UI test runner).

- Examples
  - Added path-loaded harness example:
    - `crates/no_expect_outside_tests/examples/pass_expect_in_path_module_harness.rs`
    - `crates/no_expect_outside_tests/examples/path_module_harness_support/service_tests.rs`

## Why this matters
- Previously, path-loaded modules with non-standard names could bypass the module-name checks, causing the no_expect lint to miss test contexts inside those modules. This change broadens detection rules to cover common non-standard naming patterns (e.g., `service_tests`) that arise from `#[path]`-loaded tests, ensuring the lint behavior is consistent across all test structures. It also extends recognition to rstest_parametrize-style contexts so parent test functions are properly treated as test-only.

## How to test
- Run unit tests for the driver:
  - cargo test -p no_expect_outside_tests
- Run UI/regression tests per repository guidelines (these verify that path-loaded, non-standard module names containing `#[tokio::test]` are recognized as test contexts under `--test`).
- Optionally exercise the new examples under `examples/` to reproduce the path-loaded module scenario described in the regression tests.

## Related issues
- Fixes #132

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

---
ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/6740313c-27f0-4e46-a021-2a6a9f9aca06

📝 Closes #132

## Summary by Sourcery

Broaden detection of test-only code for the no_expect_outside_tests lint, especially for rstest-based tests and non-standard, path-loaded test modules, and add regression coverage to ensure correct behavior under the --test harness.

New Features:
- Recognise additional test module naming conventions (e.g., prefix/suffix variants like service_tests) when determining test-only contexts.
- Detect rstest #[case]-generated companion modules so their parent functions are treated as test-only code.
- Support rstest_parametrize-style attributes in the shared test-like attribute registry.

Enhancements:
- Refine module name heuristics into a reusable has_test_module_name helper used by is_test_named_module.

Tests:
- Extend driver and attribute unit tests to cover new module-name and rstest_parametrize patterns.
- Add example-based regression tests for real rstest usage and non-standard tokio test modules compiled under the --test harness.
- Introduce UI fixtures validating .expect allowance in #[tokio::test] and #[cfg(test)] modules with non-standard names.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The `has_test_module_name` behavior and its tests seem inconsistent: the implementation only checks `test_`/`tests_` prefixes and `_test`/`_tests` suffixes, but `has_test_module_name_matches_test_conventions` expects "test_like_utils" to be `true`—either adjust the predicate to cover this broader pattern or change the test/example to match the documented naming rules.

## Individual Comments

### Comment 1
<location path="crates/no_expect_outside_tests/src/driver/tests.rs" line_range="139-148" />
<code_context>
+
+use rstest::rstest;
+
+#[rstest]
+#[case(1)]
+#[case(42)]
</code_context>
<issue_to_address>
**suggestion (testing):** Add more edge-case names around `"test"`/`"tests"` substrings without underscores to lock in the intended `has_test_module_name` behaviour

The new `has_test_module_name_matches_test_conventions` cases cover exact, prefix, and suffix patterns well. To better pin down the boundary and prevent future regressions, consider adding a few negative cases like:

- `"mytest"` / `"mytests"` (no underscore) → `false`
- `"testx"` / `"testsx"` → `false`
- `"service_tests_helper"` (test-like suffix with extra tail) → `false` if that’s intended

This makes it clear that arbitrary `"test"`/`"tests"` substrings should not be treated as valid test module names.
</issue_to_address>
~~~

File: crates/no_expect_outside_tests/src/driver/mod.rs [hunk 1]
```diff
@@ -319,6 +352,34 @@ fn is_matching_harness_test_descriptor(
...
+        };
+        if mod_ident.name != function_name {
+            return false;
+        }
+        // The module must contain at least one harness const descriptor to
+        // avoid false positives on non-test modules that happen to share a
+        // function name.
+        module.item_ids.iter().any(|id| {
+            let child = cx.tcx.hir_item(*id);
+            matches!(child.kind, hir::ItemKind::Const(..))
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/no_expect_outside_tests/src/driver/mod.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

~~~
<!-- This is an auto-generated reply by CodeRabbit -->
> [!TIP]
> For best results, initiate chat on the files or code changes.

<details>
<summary>🧩 Analysis chain</summary>

🏁 Script executed:
```
#!/bin/bash
echo "=== driver/tests.rs (has_test_module_name section) ==="
rg -n "has_test_module_name|test_like_utils|mytest|testx|service_tests_helper" crates/no_expect_outside_tests/src/driver/tests.rs -A2 -B2

echo ""
echo "=== driver/mod.rs (has_test_module_name implementation) ==="
rg -n "has_test_module_name|fn has_test_module" crates/no_expect_outside_tests/src/driver/mod.rs -A 20
```

Length of output: 3868

---



</details>

`@leynos`: The individual comment (edge-case negatives) has been resolved — `"mytest"`, `"mytests"`, `"testx"`, `"testsx"`, and `"service_tests_helper"` are all present with `false` expectations.

However, the **overall comment** remains unresolved and the inconsistency has actually flipped direction. In `driver/tests.rs` line 152 the case is now:

```rust
#[case::test_embedded_not_suffix("test_like_utils", false)]
```

But the implementation in `driver/mod.rs` lines 255–256 is:

```rust
|| name.starts_with("test_")
|| name.starts_with("tests_")
```

`"test_like_utils"` *does* start with `"test_"`, so `has_test_module_name("test_like_utils")` returns `true` at runtime — contradicting the test's `false` expectation. The test will fail.

Pick one of the two fixes:

**Option A — tighten the implementation** (if `test_like_utils` should not be a test module):

```rust
fn has_test_module_name(name: &str) -> bool {
    matches!(name, "test" | "tests")
        || matches!(name.split_once('_'), Some(("test" | "tests", _)))
        || name.ends_with("_test")
        || name.ends_with("_tests")
}
```

This restricts the prefix check to exactly `test_<something>` or `tests_<something>` (one `_` separator immediately after the keyword), which excludes `test_like_utils` since `split_once('_')` yields `("test", "like_utils")` — wait, that *would* match. A safer approach is to check that the remainder after the prefix contains no further `_test`/`_tests` sub-pattern, or simply accept that `test_like_utils` is a valid test-module name and go with Option B.

**Option A (revised) — require the prefix to be the whole leading segment:**

```rust
fn has_test_module_name(name: &str) -> bool {
    matches!(name, "test" | "tests")
        || name.ends_with("_test")
        || name.ends_with("_tests")
        || matches!(
            name.split_once('_'),
            Some(("test" | "tests", rest)) if !rest.contains('_')
        )
}
```

This accepts `test_helpers` but rejects `test_like_utils` (rest = `"like_utils"` contains `_`).

**Option B — align the test to the implementation** (if `test_like_utils` *should* be treated as a test module, remove the negative case and document it):

```rust
#[case::test_prefix_multi_segment("test_like_utils", true)]
```

Choose whichever matches the intended semantics; the implementation and the test must agree.
~~~
